### PR TITLE
sources: introduce new property mappings per user and group

### PR DESCRIPTION
**Pull request by rissson**
## Details

Upstream the user and group property mappings logic of the LDAP provider to all sources.

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [x] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [x] The code has been formatted (`make web`)

Documentation PR: https://github.com/goauthentik/authentik/pull/8780

### CODE REVIEW

File: authentik/core/models.py [hunk 1]
```diff
@@ -586,6 +597,76 @@ def ui_user_settings(self) -> UserSettingSerializer | None:
...
+                    EventAction.CONFIGURATION_ERROR,
+                    message=f"Failed to evaluate property mapping: '{mapping.name}'",
+                    source=self,
+                    mapping=mapping,
+                ).save()
+                LOGGER.warning("Mapping failed to evaluate", exc=exc, source=self, mapping=mapping)
+                continue
+            MERGE_LIST_UNIQUE.merge(properties, value)
+
+        return remove_none(flatten_rec(properties))
```
Comment (rissson) [inline authentik/core/models.py:668]: flatten_rec probably is a bad idea here. It should be handled in the LDAP source directly.

remove_none is a duplicate of authentik.policies.utils.delete_none, except that it is recursive.

File: authentik/sources/ldap/migrations/0004_remove_ldappropertymapping_object_field_and_more.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                root[comp] = {{}}
+            root = root.get(comp, {{}})
+        root[path_parts[-1]] = get_field()
+    else:
+        result[field] = get_field()
+    return result
+    """
+
+
+def migrate_property_mappings(apps, schema_editor):
```
[the remark this answers is not in the document]
Comment (rissson) [reply authentik/sources/ldap/migrations/0004_remove_ldappropertymapping_object_field_and_more.py]: tested with the data from our ldap source blueprint, all's good

**Comment by rissson** [comment]
This should be ready for review.

**Comment by eram** [comment]
@rissson is this coming on the next release?

File: authentik/core/models.py [hunk 3]
```diff
@@ -605,6 +617,76 @@ def ui_user_settings(self) -> UserSettingSerializer | None:
...
+        """Get base properties for a user or a group to build final properties upon."""
+        if object_type == User:
+            properties = self.get_base_user_properties(**kwargs)
+            properties.setdefault("path", self.get_user_path())
+            return properties
+        if object_type == Group:
+            return self.get_base_group_properties(**kwargs)
+        return {}
+
+    def build_object_properties(
```
[the remark this answers is not in the document]
Comment (rissson) [reply authentik/core/models.py:665]: I moved this to using propertymappingmanager where relevant. Let me know what you think.

File: authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+#     "{mapping.object_field}": ldap.get("{mapping.object_field}")
+# }}
+# Note that this example has been generated and should not be used as-is.
+def get_field():
+    {textwrap.indent(mapping.expression, prefix='    ')}
+
+field = "{mapping.object_field}"
+result = {{"attributes": {{}}}}
+if field.startswith("attributes."):
+    # Adapted from authentik/lib/config.py::set_path_in_dict
```
[the remark this answers is not in the document]
Comment (rissson) [reply authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py:29]: But I do agree that it's a mess

File: authentik/sources/ldap/sync/base.py [hunk 5]
```diff
@@ -140,62 +121,14 @@ def search_paginator(  # noqa: PLR0913
...
-            raise exc from exc
-        except PropertyMappingExpressionException as exc:
-            # Value error can be raised when assigning invalid data to an attribute
-            Event.new(
-                EventAction.CONFIGURATION_ERROR,
-                message=f"Failed to evaluate property-mapping {exception_to_string(exc)}",
-                mapping=exc.mapping,
-            ).save()
-            self._logger.warning("Mapping failed to evaluate", exc=exc, mapping=exc.mapping)
-            raise StopSync(exc, None, exc.mapping) from exc
```
Comment (rissson) [inline authentik/sources/ldap/sync/base.py:184]: We need to decide what to do with this in the general case. The way I've written it currently is that if a property mapping fails, we just ignore it.

I think it's probably better to not do that, and fail whatever is currently happening (a sync or a source login flow). Thoughts?

File: authentik/core/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (BeryJu) [reply authentik/core/models.py:673]: Move to a class next to core/source/flow_manager

File: authentik/core/models.py [hunk 6]
```diff
@@ -616,6 +631,102 @@ def ui_user_settings(self) -> UserSettingSerializer | None:
...
+                    source=self,
+                    mapping=exc.mapping,
+                ).save()
+                LOGGER.warning(
+                    "Mapping failed to evaluate",
+                    exc=exc,
+                    source=self,
+                    mapping=exc.mapping,
+                )
+                continue
```
Comment (BeryJu) [inline authentik/core/models.py:716]: Mapping exception should stop and not continue

File: authentik/core/models.py [hunk 7]
```diff
@@ -616,6 +631,102 @@ def ui_user_settings(self) -> UserSettingSerializer | None:
...
+            source=self,
+            properties=properties,
+            **kwargs,
+        )
+        while True:
+            try:
+                value, mapping = next(evaluations)
+            except StopIteration:
+                break
+            except PropertyMappingExpressionException as exc:
```
Comment (rissson) [inline authentik/core/models.py:703]: Don´t silently fail

File: authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+# Generated by Django 5.0.2 on 2024-02-29 11:21
+
+import textwrap
+
+from django.db import migrations
+
+
+def migrate_ldap_property_mappings_object_field(apps, schema_editor):
+    if schema_editor.connection.alias != "default":
+        return
```
Comment (rissson) [inline authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py:10]: Remove this, and use `using` where queries are run

File: authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    for mapping in LDAPPropertyMapping.objects.all():
+        mapping.expression = f"""
+# This property mapping has been automatically changed to
+# match the new semantics of source property mappings.
+# You can simplify it if you want.
+# You should return a dictionary of fields to set on the user or the group.
+# For instance:
+# return {{
+#     "{mapping.object_field}": ldap.get("{mapping.object_field}")
+# }}
```
Comment (BeryJu) [inline authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py:21]:
```python
# return {
#     "{mapping.object_field}": ldap.get("{mapping.object_field}")
# }
```

File: authentik/sources/ldap/sync/base.py (hunk 5 above)
Comment (rissson) [reply authentik/sources/ldap/sync/base.py:184]: done, we now fail with `PropertyMappingExpressionException` if the expression fails to evaluate

The LDAP sync code then catches this and throws StopSync, to preserve existing behaviour.

File: authentik/core/models.py (hunk 6 above)
Comment (rissson) [reply authentik/core/models.py]: done, we now fail with `PropertyMappingExpressionException` if the expression fails to evaluate

The LDAP sync code then catches this and throws StopSync, to preserve existing behaviour.

File: authentik/core/models.py (hunk 7 above)
Comment (rissson) [reply authentik/core/models.py]: done, we now fail with `PropertyMappingExpressionException` if the expression fails to evaluate

The LDAP sync code then catches this and throws StopSync, to preserve existing behaviour.

File: authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py (hunk 9 above)
Comment (rissson) [reply authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py:21]: We need that `{{` so they're not escaped by the f-string

File: authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rissson) [reply authentik/sources/ldap/migrations/0005_remove_ldappropertymapping_object_field_and_more.py]: done in 4d5a05d5365a687dcf5dfab26a4dd21ff6bc1df3

I still would like someone else to test this, just to make sure it's 100% working as intended

**Review by BeryJu — APPROVED** [review]

**Pull request merged**